### DISCUSSION TITLE
GCN-based Autoencoder not training as intended

### DISCUSSION THREAD

**Discussion by sivaramkrishnan101** [pyg-team/pytorch_geometric#6917] [Q&A]
Hi everyone, 

Dataset - I have a dataset of 10000 graphs. Each of the graph consists 3 nodes, and the nodes have a node features corresponding to their euclidean(x, y) coordinates. The graphs are fully connected. 

Problem statement - I want to train a simple graph autoencoder (GAE) and backpropagate based on loss(node_features, predicted node_features). The proposed encoder consists of 2 GCN layers, each followed by a Relu activation, very similar to the one in the examples. 

Issue - **But I have been seeing a problem, when looking at the output of the encoder part, i.e. the node embeddings, all the node embeddings have the same vector. For ex., see the image below, all I have done is generate the embedding using the GCN layer for a single graph.** 

![image](https://user-images.githubusercontent.com/108207716/225297389-2b1a6f3c-96ec-4c1f-a347-3dea98c88ae5.png)

Conclusion - I am not sure what is causing this problem, I have implemented the same GAE architecture that I have come up with on the Citeseer dataset and the results seem to be okay. But in the case of using my dataset, the training is not working fine, and it is better when I just include the FCN layers and remove the GCN layers completely. 

I would love to hear your feedback on this and reasons to why this might be happening. 

Thank you!

**Comment by rusty1s** [comment]
GCN cannot operate on a fully connected graph, it will produce identical node embeddings (that's exactly what you observe). Try replacing with `SAGEConv` :)

**Reply by sivaramkrishnan101** [reply]
Wow! Thank you, Matthias, for your prompt feedback. I will go ahead and try using the `SAGEConv` layer as you have suggested. 

P.S. Now that you mention it, I was able to figure out as to why I am getting same node embeddings for each node.

**Reply by sivaramkrishnan101** [reply]
Hi Matthias, I have tried the `SAGEConv` layer as compared to the `GCNConv`. 

I am still running into the same issues of similar node embeddings for each node (My hypothesis is that this issue arises because of the graph's small size, causing `SAGEConv` to generate such embeddings, making it ill to use SAGEConv for low-scale graphs especially when it is fully connected). 

![image](https://user-images.githubusercontent.com/108207716/225553087-bd49bf16-179b-4e77-8378-d27003fb321a.png)

I have kept the architecture of the encoder to be relatively simple just to generate some 'meaningful' embeddings before proceeding to add more layers in the architecture. 

![image](https://user-images.githubusercontent.com/108207716/225553336-719cb874-8fc6-4a58-8ec8-5c90e71dd1fb.png)

I plan to experiment with other GNN layers available in the `torch.nn` module mentioned here: https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html. 

**Kindly inform me if you have any suggestions.**

**Reply by rusty1s** [reply]
I think `SAGEConv` will only produce identical embeddings if all your input features are identical as well. Is that the case?

**Reply by sivaramkrishnan101** [reply]
Hello Matthias,

I appreciate your response.

I made mistakes in my previous attempts, but I have corrected them now. In short, I am now capable of creating distinct node embeddings for every node.

Thank you very much for your help.

![image](https://user-images.githubusercontent.com/108207716/225564852-e5a0716e-8b69-4073-bd9e-47e0e6b5e5f4.png)
This is what it looks like after using just a single layer of `SAGEConv` (I have not used any activations yet).
